src/installer/corehost/cli/apphost/bundle/extractor.cpp

```diff
@@ -126,58 +113,138 @@ void extractor_t::begin()
+        throw StatusCode::BundleExtractionFailure;
+    }
 
-            dir_utils_t::remove_directory_tree(m_working_extraction_dir);
-            break;
-        }
+    trace::info(_X("Completed new extraction."));
+}
 
-        if (should_retry && (retry_count--) > 0)
-        {
-            trace::info(_X("Retrying extraction due to EACCES trying to rename the extraction folder to [%s]."), m_extraction_dir.c_str());
-            pal::sleep(100);
-            continue;
-        }
-        else
-        {
-            trace::error(_X("Failure processing application bundle."));
-            trace::error(_X("Failed to commit extracted files to directory [%s]."), m_extraction_dir.c_str());
-            throw StatusCode::BundleExtractionFailure;
-        }
+void extractor_t::commit_file(const pal::string_t& relative_path)
+{
+    // Commit individual files to the final extraction directory.
+
+    pal::string_t working_file_path = working_extraction_dir();
+    append_path(&working_file_path, relative_path.c_str());
+
+    pal::string_t final_file_path = extraction_dir();
+    append_path(&final_file_path, relative_path.c_str());
+
+    if (dir_utils_t::has_dirs_in_path(relative_path))
+    {
+        dir_utils_t::create_directory_tree(get_directory(final_file_path));
     }
+
+    bool extracted_by_concurrent_process = false;
+    bool extracted_by_current_process =
+        dir_utils_t::rename_with_retries(working_file_path, final_file_path, extracted_by_concurrent_process);
+
+    if (extracted_by_concurrent_process)
+    {
+        // Another process successfully extracted the dependencies
+        trace::info(_X("Extraction completed by another process, aborting current extraction."));
+    }
+
+    if (!extracted_by_current_process && !extracted_by_concurrent_process)
+    {
+        trace::error(_X("Failure processing application bundle."));
+        trace::error(_X("Failed to commit extracted files to directory [%s]."), extraction_dir().c_str());
+        throw StatusCode::BundleExtractionFailure;
+    }
+
+    trace::info(_X("Extraction recovered [%s]"), relative_path.c_str());
 }
 
-void extractor_t::extract(const manifest_t& manifest, reader_t& reader)
+void extractor_t::extract_new(reader_t& reader)
 {
     begin();
-    for (const file_entry_t& entry : manifest.files) {
+    for (const file_entry_t& entry : m_manifest.files) 
+    {
         extract(entry, reader);
     }
-    commit();
+    commit_dir();
+}
+
+// Verify an existing extraction contains all files listed in the bundle manifest.
+// If some files are missing, extract them individually.
+void extractor_t::verify_recover_extraction(reader_t& reader)
+{
+    pal::string_t& ext_dir = extraction_dir();
+    bool recovered = false;
+
+    for (const file_entry_t& entry : m_manifest.files)
+    {
+        pal::string_t file_path = ext_dir;
+        append_path(&file_path, entry.relative_path().c_str());
+
+        if (!pal::file_exists(file_path))
```

Comment:
I experimented using `readdir()`  /  `FindFirstFile()/FindNextFile()` APIs in place of `file_exists()` checks on each file.
That is, create a dictionary of files in the the extraction location (and its subdirectories), and check that it has all entries in the manifest.

However, there wasn't a significant difference in startup time due to the change -- especially since the startup impact is already small. Given this and the fact that .net 5 will focus on moving away from extraction, I kept the file_exists checks to keep the logic simple.